### ISSUE TITLE
tasks and discussion for new GPU CI queue

### ISSUE THREAD

**Issue by beckermr** [conda-forge/conda-forge.github.io#1272]
This issue is to document tasks and todos for the new GPU CI queue.

To-Do Items:
 - [x] decide on software to provision the CI (decided on drone)
   - best candidates from internal discussions are drone or azure 
 - [ ] develop a process to permission feedstocks on the CI
   - [ ] we will want an allow list of feedstocks that are allowed to use the queue
   - [ ] we will want to add a job to conda-forge/admin-requests to add feedstocks to this list and provision them with the proper keys/permissions to access the CI
   - [ ] add allowed list of users
 - [ ] put in changes to smithy to allow separate build and test phases in the CI config files
 - [ ] make sure the build phase does not tie up a GPU in the CI system
   - [ ] separate queues for build and test
   - [ ] move jobs from build on cpu to test on GPU
 - [ ] put in monitoring for the load on the queues
   - we have existing tools that will be able to output the load in five-minute increments to the conda-forge status page
   - we may want more than this however
 - [ ] establish and document a clear process of who to contact when things fail or break

cc @dharhas @jakirkham @kkraus14 @viniciusdc

@mariusvniekerk for azure stuff

I missed Ray and Kent on the github handles. Can someone ping them here?

hackmd: https://hackmd.io/QCX9xMnzS2WeobW0athINA

**Comment by jakirkham** [comment]
cc @ocefpaf @mike-wendt @raydouglass @teoliphant

**Comment by jakirkham** [comment]
cc @h-vetinari (Axel)

**Comment by jakirkham** [comment]
Should add I also don't know Kent's or others' GH handles. So please cc others as needed. Thanks! 😄

**Comment by beckermr** [comment]
Ahhhh thanks! I could not find Axel's github handle. See this one too for the other azure stuff: https://github.com/conda-forge/conda-forge.github.io/issues/1273

**Comment by dharhas** [comment]
cc @Kllewelyn @rtiwariops from openteams.

**Comment by leofang** [comment]
Very exciting!

**Comment by h-vetinari** [comment]
Thanks for opening this @beckermr!

Linking my [closing comment](https://github.com/conda-forge/conda-forge.github.io/issues/1062#issuecomment-798506357) from #1062 for reference. TL;DR:
> In short: it would be amazing if any of the involved people / companies could take this as impetus to chip in something as well. GPU computing is only ever going to get larger, and I believe that sharing some (comparatively low) CI infra costs would provide huge bang-per-buck for the people & companies building & using such packages.

**Comment by dharhas** [comment]
Adding @leej3 from Quansight.

**Comment by jakirkham** [comment]
cc @aktech (as it looks like you have been doing work in this area as well)

**Comment by viniciusdc** [comment]
@beckermr have we decided upon Azure as the software to provision the CI for the GPU tasks?

**Comment by dharhas** [comment]
Hey Folks,

Looks like this is finally happening. We expect hardware to arrive in 2-4 weeks. There are still a lot of questions that are unanswered about the software stack we should run on it and getting this setup and managed etc. So I wanted to restart the discussion. 

Also adding @jaimergp to the conversation.

**Comment by jakirkham** [comment]
Would it make sense to have a meeting?

**Comment by dharhas** [comment]
I think a coordination meeting makes sense. We probably need some higher bandwidth time to get broad strokes of what this will loof like sorted out.

**Comment by jakirkham** [comment]
Ok. Went ahead and created [a poll]( https://whenisgood.net/ehwbe8j ) for us to figure out when is best time to meet in the next 2 weeks. Also please make sure to configure timezone before filling out the poll. Will share the results here and we can go from there

**Comment by leofang** [comment]
May I invite myself? 😛

**Comment by jakirkham** [comment]
Appreciate the general enthusiasm around this work! 😄

Anyone is welcome. Though my guess is this will be focused on technical issues around integration into conda-forge. So doubt this will be of interest outside those planning to do that work. That said, we can take notes, raise new issues, and summarize here for broader community awareness.

**Comment by h-vetinari** [comment]
Awesome news! Would love to participate, but on holidays for the next two weeks 😅

**Comment by beckermr** [comment]
hey @jakirkham! I totally missed this poll. If there is a time for the next meeting already, that is fine. Otherwise, I have filled out the poll.

**Comment by jakirkham** [comment]
Ok every time has some conflicts for someone. That said, the least conflicting time is on 27 July at 9a US Pacific / 11a US Central / 12p US Eastern / 5p UK / 6p European. We can take notes and summarize here for those that miss. Will send out an invite and we can go from there.

**Comment by jakirkham** [comment]
Alright have sent that out 📬

Think I got everyone who responded to the poll. Though feel free to forward to others that I may have been missed.

Also set it up with Microsoft Teams since that's what I have easy access to. Though if people prefer to use something different, feel free to propose (and be ready to set it up 😉). Otherwise we will stick with Teams.

Thanks all! 😄

**Comment by viniciusdc** [comment]
I'm not totally sure about the feasibility of this, but Drone seems to have and [admin](https://docs.drone.io/server/user/admin/) management for queues

**Comment by dharhas** [comment]
Just noticed that drone's open source version is fairly hobbled vs their paid version.

https://www.drone.io/enterprise/opensource/#features

Not sure if we need any of the features that are not present in the OSS version but I thought I'd raise.

**Comment by beckermr** [comment]
Thanks for this. We'll have to find out by doing I imagine.

**Comment by leofang** [comment]
btw What's the GPU model that the CI would use? I was under the wrong impression yesterday that MIG would work out of box for any existing models, but it looks like only certain Ampere GPUs support this feature.

**Comment by viniciusdc** [comment]
@beckermr about the current status of the CI-run integration, do you think we can have a feedstock to run some tests and  heck how the permissioning will be held?

**Comment by beckermr** [comment]
I do not.

**Comment by viniciusdc** [comment]
> I do not.

Should we add a test suit somewhere in the bots tests to eval the GPU builds before enabling it? I am open to any suggestions to test this integration.

**Comment by beckermr** [comment]
We need to hear back from the ci run folks.

**Comment by jaimergp** [comment]
Will do!

**Comment by leofang** [comment]
Just curious, what's the status here?

**Comment by jaimergp** [comment]
Hi Leo!

I gave a little update in the last call, but I should have posted it here too, my apologies!

@viniciusdc and I are working on a test feedstock (outside of conda-forge while prototyping) to build OpenMM on the GPU server. We can successfully run CPU builds with almost no changes in the conda-forge machinery (same Docker images and scripts). We are now debugging the GPU drivers, which are preventing us from correctly invoking the nvidia-docker integrations.

We are now waiting for the datacenter tech support to get back to us on the ticket we submitted some days ago.

So, slowly but surely we are getting there!
